README.md
# drag_like
1. Flutter左滑右滑/喜欢、不喜欢的特效插件
2. 支持划出事件回调，通知上层是左边划出还是右边划出
3. 支持滑动进度回调
4. 支持第二层widget缩放完成回调，缩放完成时刷新数据
5. 支持手指抬起回调
6. 支持手速检测，速度超过指定值后，卡片将会划出屏幕，已解决手速不准确问题（感谢[@DoodleBears](https://github.com/DoodleBears)贡献代码）
7. null safety
8. 支持控制控制左右划出
9. 支持设置滑出动画执行时长
10. 支持控制器左右划出时，自定义回调参数

### 安装
在工程 pubspec.yaml 中加入 dependencies

```
dependencies:
  drag_like: ^last version
```
## 效果图
<img src="https://user-images.githubusercontent.com/21274377/134722410-d38307ff-4506-47b2-9e40-ca64545a15f4.gif" width="400">

## 使用方法
```
DragLike(
  dragController: _dragController,
  duration: Duration(milliseconds: 520),
  child: imagelist.length <=0 ? Text('加载中...') : ClipRRect(
    borderRadius: BorderRadius.circular(10),
    child: TestDrag(src:imagelist[0])
  ), 
  secondChild: imagelist.length <= 1 ? Container(): ClipRRect(
    borderRadius: BorderRadius.circular(10),
    child:TestDrag(src:imagelist[1])
  ), 
  screenWidth: 375, 
  outValue: 0.6,
  dragSpeedRatio: 80,
  onChangeDragDistance: (distance){
    /// {distance: 0.17511112467447917, distanceProgress: 0.2918518744574653}
    print(distance.toString());
  },
  onOutComplete: (type){
    /// left or right
    print(type);
  },
  onScaleComplete: (){
    imagelist.remove(imagelist[0]);
    if(imagelist.length == 0) {
      loaddata();
    }
    setState(() {});
  },
  onPointerUp: (){
    
  },
),
```

## 回调方法
### onOutComplete
```
onOutComplete: (type){
	/// left or right
	print(type);
},
```

### onScaleComplete
```
onScaleComplete: (){
	imagelist.remove(imagelist[0]);
	if(imagelist.length == 0) {
		loaddata();
	}
	setState(() {});
},
```

### onChangeDragDistance
1. distance滑动距离，0-1，超过设置的值将会划出widget
2. distanceProgress，根据设置的划出边界值的百分比，0-1+，方便做一些特效，超过1即代表松手就会划出边界
```

onChangeDragDistance: (distance){
	/// {distance: 0.17511112467447917, distanceProgress: 0.2918518744574653}
	print(distance.toString());
},
```
### onPointerUp
```
onPointerUp: (){

},
```

## 控制器
```

_dragController.toLeft(completeTag: 'custom_left');

_dragController.toRight(completeTag: 'custom_right');

```

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [2.1.1]

* 增加控制器控制左滑右滑时，自定义回调参数，默认回调left、right

## [2.1.0]

* 增加控制器dragController，可以直接控制左滑右滑
* 增加卡片划出时长参数：duration

## [2.0.21]

* 优化惯性滑动，解决手抬起时的假惯性问题

## [2.0.12]

* update readme

## [2.0.11]

* update readme

## [2.0.1]

* 性能优化

## [2.0.0]

*  null-safety

## [1.0.4]

*  增加卡片划出回调

## [1.0.3]

*  部分代码优化

## [1.0.2]

*  支持手速检测，速度超过指定值后，卡片将会划出屏幕

## [1.0.1]

*  优化手指抬起判断

## [1.0.0]

*  插件init


test/drag_like_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('drag_like');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    // expect(await DragLike.platformVersion, '42');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:drag_like_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/drag_like.dart
import 'package:flutter/material.dart';

class DragLike extends StatefulWidget {
  final Widget child;
  final Widget secondChild;
  final double screenWidth;
  final double outValue;
  final double dragSpeed;
  final Duration ? duration;
  final ValueChanged<String?> onOutComplete;
  final VoidCallback onScaleComplete;
  final ValueChanged<Map> onChangeDragDistance;
  final VoidCallback onPointerUp;
  final DragController ? dragController;
  DragLike(
      {Key ? key,
      required this.child,
      required this.secondChild,
      required this.screenWidth,
      required this.outValue,
      required this.dragSpeed,
      required this.onOutComplete,
      required this.onScaleComplete,
      required this.onChangeDragDistance,
      required this.onPointerUp, 
      this.dragController, 
      this.duration})
      : assert(outValue <= 1 && outValue > 0),
        assert(dragSpeed > 0),
        super(key: key);

  @override
  _DragLikeState createState() => _DragLikeState();
}

class _DragLikeState extends State<DragLike> with TickerProviderStateMixin {
  // 滑动不到指定位置返回时的动画
  Animation ? animation;
  // 第二个页面动画到前面
  Animation ? scaleAnimation;
  // 按下时的X坐标，以此判断是左滑还是右滑
  double onPressDx = 0;
  // 拖拽时，上一瞬间x轴的位置
  double lastPositionX = 0;
  // 屏幕宽度
  double get screenWidth => widget.screenWidth;
  // 旋转角度
  // double angle = 0;
  final ValueNotifier<double> angle = ValueNotifier(0);
  // 旋转时，x轴的偏移量
  double offsetX = 0;
  // 拖拽时，两个瞬间之间，x轴的差值
  double distanceBetweenTwo = 0;
  // 拖拽发生的时间
  DateTime dragTime = DateTime(0);
  // 第二层的缩放值，0-0.1，因为已设置初始值为0.9
  final ValueNotifier<double> secondScale = ValueNotifier(0);
  // 拖拽距离比0.0-1
  double dragDistance = 0;
  // 滑动流畅值，默认3.8，越小越快
  double dragGvalue = 5;
  // 第二层缩放速度，默认4，越小越快
  double secondScaleSd = 2.3;
  // 动画是否执行完成，执行完成后，控制器执行滑动才有效
  bool scaleComplete = true;

  @override
  void initState() {
    super.initState();
    widget.dragController?.controlerInit(this);
  }

  // 更新偏移和缩放量
  void updatePosition(positionX,{iscontroller = false}) {
    // print("positionX = " + positionX.toString());
    // print("lastPositionX = " + lastPositionX.toString());
    double offset = positionX - onPressDx;
    // print("offset = " + offset.toString());

    dragTime = DateTime.now();

    distanceBetweenTwo = positionX - lastPositionX;
    lastPositionX = positionX;

    double offsetAbs = offset.abs();
    // print("offsetAbs = " + offsetAbs.toString());
    double angleTemp =
        double.parse((offset / screenWidth / dragGvalue).toStringAsFixed(3));
    if (angle.value != angleTemp) {
      angle.value = angleTemp;
      secondScale.value = (offsetAbs / screenWidth / dragGvalue) / secondScaleSd;
      if (secondScale.value >= 0.1) secondScale.value = 0.1;
      dragDistance = offsetAbs / screenWidth;
      // print("dragDistance = " + dragDistance.toString());

      if (offset < 0) {
        offsetX = -80;
      } else {
        offsetX = 80;
      }

      double distance = offset / screenWidth;
      double distanceProgress = distance / widget.outValue;
      if(!iscontroller) {
        widget.onChangeDragDistance({
          'distance': distance,
          'distanceProgress': distanceProgress.abs(),
        });
      }
    }
  }

  // 上层以及第二层返回动画执行
  runBackAnimate() {
    AnimationController controller;
    controller = AnimationController(
        duration: const Duration(milliseconds: 250), vsync: this);
    this.animation = Tween<double>(begin: angle.value, end: 0).animate(controller)
      ..addListener(() {
        angle.value = this.animation!.value;
        secondScale.value = angle.value.abs() / secondScaleSd;
        if (secondScale.value <= 0) secondScale.value = 0;

        dragDistance = 0;
        if (controller.isCompleted) {
          controller.dispose();
        }
        // setState(() {});
      });
    controller.forward(from: 0);
  }

  void runOutAnimate(int type,{iscontroller = false,String completeTag=''}) {
    scaleComplete = false;
    // 回调给上层
    widget.onOutComplete(type == 1 ? (completeTag == '' ? 'right' : completeTag) : (completeTag == '' ? 'left' : completeTag));

    AnimationController controller;
    controller = AnimationController(
        duration: widget.duration == null ? const Duration(milliseconds: 350) : widget.duration, vsync: this);
    // 如果是控制器直接控制滑动，旋转角度直接从0开始
    if(iscontroller) angle.value = 0;
    this.animation = Tween<double>(begin: angle.value, end: type == 1 ? 0.5 : -0.5)
      .animate(controller)
        ..addListener(() {
          angle.value = this.animation!.value;
          dragDistance = 0;
          if (controller.isCompleted) {
            controller.dispose();
          }
          // setState(() {});
        });
    
    controller.forward(from: 0);
    if(iscontroller) {
      secondScale.value = 0.0;
    }
  }

  void runInScaleAnimate() async {
    AnimationController controller;
    controller = AnimationController(
        duration: widget.duration == null ? const Duration(milliseconds: 350) : widget.duration, vsync: this);
    this.scaleAnimation =
        Tween<double>(begin: secondScale.value, end: 0.1).animate(controller)
          ..addListener(() async {
            secondScale.value = this.scaleAnimation!.value;
            if (controller.isCompleted) {
              widget.onScaleComplete();
              scaleComplete = true;
              controller.dispose();
              // 缩放完成以后，将上一层的widget还原到起始位置，不要动画，业务方需要将上层widget的数据替换
              angle.value = 0;
            }
            // setState(() {});
          });
    controller.forward(from: 0);
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
        child: Stack(
          children: [
            ValueListenableBuilder(
              valueListenable: secondScale,
              builder: (BuildContext context,double value,Widget ? child){
                return Transform.scale(
                  scale: 0.9 + secondScale.value,
                  child: child,
                );
              },
              child: widget.secondChild,
            ),
            GestureDetector(
              onPanDown: (DragDownDetails value) {
                lastPositionX = onPressDx = value.globalPosition.dx;
              },
              onHorizontalDragUpdate: (DragUpdateDetails value){
                updatePosition(value.globalPosition.dx);
              },
              onHorizontalDragEnd: (DragEndDetails details){
                // 滑动速度
                var dragSpeed = details.velocity.pixelsPerSecond.dx.abs();
                if (dragDistance > widget.outValue || dragSpeed >= widget.dragSpeed) {

                  if(distanceBetweenTwo > 0) {
                    // right
                    runOutAnimate(1);
                  } else {
                    // left
                    runOutAnimate(-1); 
                  }
                  runInScaleAnimate();
                } else {
                  runBackAnimate();
                }
                // 手指抬起时，回调给上层
                widget.onPointerUp();

              },
              child: ValueListenableBuilder(
                valueListenable: angle,
                builder: (BuildContext context,double value,Widget ? child){
                  return Transform.rotate(
                    angle: value,
                    origin: Offset(value + offsetX, 1500),
                    alignment: Alignment.lerp(Alignment.center, Alignment.bottomCenter, 1),
                    child: child,
                  );
                },
                child: widget.child,
              ),
            )
        ],
      )
    );
  }
}


class DragController {
  _DragLikeState ?  _widgetState;
  void controlerInit(_DragLikeState _widgetState){
    this._widgetState = _widgetState;
  }


  // 通过控制器左滑出以后，默认回调给上层left，这里可以根据业务自定义把left修改为custom_left等等
  void toLeft({String completeTag = ''}) {
    if(_widgetState!.scaleComplete) {
      _widgetState!.updatePosition(0.0,iscontroller: true);
      _widgetState!.runOutAnimate(-1,iscontroller: true,completeTag: completeTag);
      _widgetState!.runInScaleAnimate();
    }
  }

  // 通过控制器右滑出以后，默认回调给上层right，这里可以根据业务自定义把right修改为custom_right等等
  void toRight({String completeTag = ''}) {
    if(_widgetState!.scaleComplete) {
      _widgetState!.updatePosition(0.0,iscontroller:true);
      _widgetState!.runOutAnimate(1,iscontroller: true,completeTag:completeTag);
      _widgetState!.runInScaleAnimate();
    }
  }

}


lib/drag_like copy 2.dart
import 'package:flutter/material.dart';

class DragLike extends StatefulWidget {
  final Widget child;
  final Widget secondChild;
  final double screenWidth;
  final double outValue;
  final double dragSpeedRatio;
  final ValueChanged<String?> onOutComplete;
  final VoidCallback onScaleComplete;
  final ValueChanged<Map> onChangeDragDistance;
  final VoidCallback onPointerUp;
  DragLike(
      {Key ? key,
      required this.child,
      required this.secondChild,
      required this.screenWidth,
      required this.outValue,
      required this.dragSpeedRatio,
      required this.onOutComplete,
      required this.onScaleComplete,
      required this.onChangeDragDistance,
      required this.onPointerUp})
      : assert(outValue <= 1 && outValue > 0),
        assert(dragSpeedRatio > 0),
        super(key: key);

  @override
  _DragLikeState createState() => _DragLikeState();
}

class _DragLikeState extends State<DragLike> with TickerProviderStateMixin {
  // 滑动不到指定位置返回时的动画
  Animation ? animation;
  // 第二个页面动画到前面
  Animation ? scaleAnimation;
  // 按下时的X坐标，以此判断是左滑还是右滑
  double onPressDx = 0;
  // 拖拽时，上一瞬间x轴的位置
  double lastPositionX = 0;
  // 屏幕宽度
  double get screenWidth => widget.screenWidth;
  // 旋转角度
  // double angle = 0;
  final ValueNotifier<double> angle = ValueNotifier(0);
  // 旋转时，x轴的偏移量
  double offsetX = 0;
  // 拖拽时，两个瞬间之间，x轴的差值
  double distanceBetweenTwo = 0;
  // 拖拽发生的时间
  DateTime dragTime = DateTime(0);
  // 第二层的缩放值，0-0.1，因为已设置初始值为0.9
  // double secondScale = 0;
  final ValueNotifier<double> secondScale = ValueNotifier(0);
  // 拖拽距离比0.0-1
  double dragDistance = 0;
  // 滑动流畅值，默认3.8，越小越快
  double dragGvalue = 5;
  // 第二层缩放速度，默认4，越小越快
  double secondScaleSd = 2.3;

  @override
  void initState() {
    super.initState();
  }

  // 更新偏移和缩放量
  void updatePosition(positionX) {
    // print("positionX = " + positionX.toString());
    // print("lastPositionX = " + lastPositionX.toString());
    double offset = positionX - onPressDx;
    // print("offset = " + offset.toString());

    dragTime = DateTime.now();

    distanceBetweenTwo = positionX - lastPositionX;
    // print("distanceBetweenTwo = " + distanceBetweenTwo.toString());
    lastPositionX = positionX;

    double offsetAbs = offset.abs();
    // print("offsetAbs = " + offsetAbs.toString());
    double angleTemp =
        double.parse((offset / screenWidth / dragGvalue).toStringAsFixed(3));
    if (angle.value != angleTemp) {
      angle.value = angleTemp;
      secondScale.value = (offsetAbs / screenWidth / dragGvalue) / secondScaleSd;
      if (secondScale.value >= 0.1) secondScale.value = 0.1;
      dragDistance = offsetAbs / screenWidth;
      // print("dragDistance = " + dragDistance.toString());

      if (offset < 0) {
        offsetX = -80;
      } else {
        offsetX = 80;
      }

      double distance = offset / screenWidth;
      double distanceProgress = distance / widget.outValue;
      widget.onChangeDragDistance({
        'distance': distance,
        'distanceProgress': distanceProgress.abs(),
      });
      // setState(() {});
    }
  }

  // 上层以及第二层返回动画执行
  runBackAnimate() {
    AnimationController controller;
    controller = AnimationController(
        duration: const Duration(milliseconds: 250), vsync: this);
    this.animation = Tween<double>(begin: angle.value, end: 0).animate(controller)
      ..addListener(() {
        angle.value = this.animation!.value;
        secondScale.value = angle.value.abs() / secondScaleSd;
        if (secondScale.value <= 0) secondScale.value = 0;

        dragDistance = 0;
        if (controller.isCompleted) {
          controller.dispose();
        }
        // setState(() {});
      });
    controller.forward(from: 0);
  }

  void runOutAnimate(int type) {
    AnimationController controller;
    controller = AnimationController(
        duration: const Duration(milliseconds: 350), vsync: this);
    // direction: 判断卡片飞出方向的依据
    double direction = distanceBetweenTwo;
    if (type == 1) direction = angle.value;
    this.animation =
        Tween<double>(begin: angle.value, end: direction > 0 ? 0.5 : -0.5)
            .animate(controller)
              ..addListener(() {
                angle.value = this.animation!.value;

                dragDistance = 0;
                if (controller.isCompleted) {
                  controller.dispose();
                }
                // setState(() {});
              });
    controller.forward(from: 0);
  }

  void runInScaleAnimate() async {
    AnimationController controller;
    controller = AnimationController(
        duration: const Duration(milliseconds: 350), vsync: this);
    this.scaleAnimation =
        Tween<double>(begin: secondScale.value, end: 0.1).animate(controller)
          ..addListener(() async {
            secondScale.value = this.scaleAnimation!.value;
            if (controller.isCompleted) {
              widget.onScaleComplete();
              controller.dispose();
              // 缩放完成以后，将上一层的widget还原到起始位置，不要动画，业务方需要将上层widget的数据替换
              angle.value = 0;
            }
            // setState(() {});
          });
    controller.forward(from: 0);
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
        child: Stack(
          children: [
            ValueListenableBuilder(
              valueListenable: secondScale,
              builder: (BuildContext context,double value,Widget ? child){
                return Transform.scale(
                  scale: 0.9 + secondScale.value,
                  child: child,
                );
              },
              child: widget.secondChild,
            ),
            Listener(
              onPointerDown: (value) {
                onPressDx = value.position.dx;
                lastPositionX = onPressDx;
                // print("--------------------------------");
              },
              onPointerMove: (value) {
                updatePosition(value.position.dx);
              },
              onPointerUp: (value) async {
                int timeOffset = DateTime.now().difference(dragTime).inMicroseconds;
                //滑动速率 = 两个瞬间的距离 / 两个瞬间的时间差
                // print("timeOffset = " + timeOffset.toString());
                if (timeOffset < 5000)
                  timeOffset = 7500;
                else if (timeOffset < 8000) timeOffset = 10000;
                double dragSpeed = (distanceBetweenTwo / timeOffset * 1e5).abs();

                // print("distanceBetweenTwo = " + distanceBetweenTwo.toString());
                // print("dragSpeed = " + dragSpeed.toString());
                if (dragDistance > widget.outValue || dragSpeed >= widget.dragSpeedRatio) {
                    widget.onOutComplete(offsetX > 0 ? 'right' : 'left');
                  if (dragDistance > widget.outValue) {
                    //以angle是否达到出界angle判断
                    runOutAnimate(1); 
                    // print(offsetX > 0 ? 'right' : 'left');
                    // print("type: outValue");
                  } else {
                    //以两个瞬间滑动的方向来判断
                    runOutAnimate(-1); 
                    // print(distanceBetweenTwo > 0 ? 'right' : 'left');
                    // print("type: speed direction");
                  }
                  runInScaleAnimate();
                } else {
                  runBackAnimate();
                }
                // 手指抬起时，回调给上层
                widget.onPointerUp();
              },
              child: ValueListenableBuilder(
                valueListenable: angle,
                builder: (BuildContext context,double value,Widget ? child){
                  return Transform.rotate(
                    angle: value,
                    origin: Offset(value + offsetX, 1500),
                    alignment:
                        Alignment.lerp(Alignment.center, Alignment.bottomCenter, 1),
                    child: child,
                  );
                },
                child: widget.child,
              ),
            )
        ],
      )
    );
  }
}


android/src/main/java/com/draglike/drag_like/DragLikePlugin.java
package com.draglike.drag_like;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** DragLikePlugin */
public class DragLikePlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "drag_like");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


example/lib/main.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:drag_like/drag_like.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  DragController _dragController = DragController();
  List data = [
    'https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fn.sinaimg.cn%2Fsinacn20%2F170%2Fw1024h1546%2F20180318%2Fa00d-fyshfur3814572.jpg&refer=http%3A%2F%2Fn.sinaimg.cn&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1613903060&t=651082a2ee0be03315c114381adaea8dhttps://gimg2.baidu.com/image_search/src=http%3A%2F%2Fn.sinaimg.cn%2Fsinacn20%2F170%2Fw1024h1546%2F20180318%2Fa00d-fyshfur3814572.jpg&refer=http%3A%2F%2Fn.sinaimg.cn&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1613903060&t=651082a2ee0be03315c114381adaea8d',
    'https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fn1.itc.cn%2Fimg8%2Fwb%2Frecom%2F2016%2F05%2F19%2F146364216575228138.JPEG&refer=http%3A%2F%2Fn1.itc.cn&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1614001788&t=3727c8bd4ef9b45d3749fa42a6f28081',
    'https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fb-ssl.duitang.com%2Fuploads%2Fitem%2F201302%2F17%2F20130217172444_rKtvc.jpeg&refer=http%3A%2F%2Fb-ssl.duitang.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1614001866&t=a54e01b85f08ceaff5511f8def64ef38',
    'https://ss0.bdstatic.com/70cFuHSh_Q1YnxGkpoWK1HF6hhy/it/u=1220965489,466788603&fm=26&gp=0.jpg',
    'https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fp4.qhimg.com%2Ft01613ae824dd1edf73.jpg&refer=http%3A%2F%2Fp4.qhimg.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1614001866&t=4eee5185802274b50430200acd6fcbec',
    'https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fpic.jj20.com%2Fup%2Fallimg%2Fmn01%2F022319225536%2F1Z223225536-4.jpg&refer=http%3A%2F%2Fpic.jj20.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1614001866&t=dc5b44cf1f3581c1a787d1e4673ebc57',
    'https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fpic1.win4000.com%2Fwallpaper%2F2017-12-11%2F5a2e3dc01ae8c.jpg&refer=http%3A%2F%2Fpic1.win4000.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1614001866&t=f2eaf54e9c947e66a4ac87ccc4faa8dc',
    'https://gimg2.baidu.com/image_search/src=http%3A%2F%2Fphoto.meifajie.com%2Fpictures%2F2018-05%2F180507_152555_31502.jpg&refer=http%3A%2F%2Fphoto.meifajie.com&app=2002&size=f9999,10000&q=a80&n=0&g=0n&fmt=jpeg?sec=1614001866&t=7ffe32272026d19d863d4566ba675f31',
  ];

  List imagelist = [];

  void loaddata() async {
    await Future.delayed(Duration(milliseconds: 2000));
    imagelist.addAll(data);
    setState(() {});
  }

  @override
  void initState() {
    super.initState();
    loaddata();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('DragLike Example'),
        ),
        body: Center(
          child: Container(
              width: double.infinity,
              height: double.infinity,
              child: Stack(
                children: [
                  Container(
                    padding: EdgeInsets.only(left: 10, right: 10, top: 10, bottom: 10),
                    child: DragLike(
                      dragController: _dragController,
                      duration: Duration(milliseconds: 520),
                      child: imagelist.length <= 0
                          ? Text('加载中...')
                          : ClipRRect(
                              borderRadius: BorderRadius.circular(10),
                              child: TestDrag(src: imagelist[0])),
                      secondChild: imagelist.length <= 1
                          ? Container()
                          : ClipRRect(
                              borderRadius: BorderRadius.circular(10),
                              child: TestDrag(src: imagelist[1])),
                      screenWidth: 375,
                      outValue: 0.8,
                      dragSpeed: 1000,
                      onChangeDragDistance: (distance) {
                        /// {distance: 0.17511112467447917, distanceProgress: 0.2918518744574653}
                        // print(distance.toString());
                      },
                      onOutComplete: (type) {
                        /// left or right
                        print(type);
                      },
                      onScaleComplete: () {
                        imagelist.remove(imagelist[0]);
                        if (imagelist.length == 0) {
                          loaddata();
                        }
                        setState(() {});
                      },
                      onPointerUp: () {},
                    ),
                  ),
                  Positioned(
                    left: 0,
                    bottom: MediaQueryData.fromWindow(window).padding.bottom + 50,
                    child: Container(
                      width: MediaQueryData.fromWindow(window).size.width,
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceAround,
                        children: [
                          ElevatedButton(
                            style: ButtonStyle(
                              backgroundColor: MaterialStateProperty.all(
                                Colors.blueAccent
                              ),
                              elevation: MaterialStateProperty.all(0),
                              shape: MaterialStateProperty.all(
                                RoundedRectangleBorder(
                                  borderRadius: BorderRadius.circular(25),
                                  side: BorderSide(color: Colors.transparent),
                                )
                              ),
                              padding: MaterialStateProperty.all(
                               EdgeInsets.all(20)
                              ),
                            ),
                            child: Text('Left',style: TextStyle(color: Colors.white,fontWeight: FontWeight.w600,fontSize: 30),),
                            onPressed: () async{
                              if(imagelist.length > 0) _dragController.toLeft(completeTag: 'custom_left');
                            }
                          ),

                          ElevatedButton(
                            style: ButtonStyle(
                              backgroundColor: MaterialStateProperty.all(
                                Colors.pinkAccent
                              ),
                              elevation: MaterialStateProperty.all(0),
                              shape: MaterialStateProperty.all(
                                RoundedRectangleBorder(
                                  borderRadius: BorderRadius.circular(25),
                                  side: BorderSide(color: Colors.transparent),
                                )
                              ),
                              padding: MaterialStateProperty.all(
                               EdgeInsets.all(20)
                              ),
                            ),
                            child: Text('right',style: TextStyle(color: Colors.white,fontWeight: FontWeight.w600,fontSize: 30),),
                            onPressed: () async{
                              if(imagelist.length > 0) _dragController.toRight(completeTag: 'custom_right');
                            }
                          ),
                        ],
                      ),
                    )
                  )
                ],
              )),
        ),
      ),
    );
  }
}

class TestDrag extends StatefulWidget {
  final String src;
  const TestDrag({Key ? key,required this.src}) : super(key: key);

  @override
  _TestDragState createState() => _TestDragState();
}

class _TestDragState extends State<TestDrag> {
  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      height: double.infinity,
      color: Colors.grey,
      child: Stack(
        children: [
          Positioned.fill(child: Image.network(
            widget.src,
            fit: BoxFit.cover,
          )),
          // Text(widget.src)
        ],
      ),
    );
  }
}

// class TestDrag extends StatelessWidget {
//   final String src;
//   const TestDrag({Key ? key,required this.src}) : super(key: key);

//   @override
//   Widget build(BuildContext context) {
//     print(123333);
//     return Container(
//       width: double.infinity,
//       height: double.infinity,
//       color: Colors.grey,
//       child: Stack(
//         children: [
//           Positioned.fill(child: Image.network(
//             src,
//             fit: BoxFit.cover,
//           )),
//           Text(src)
//         ],
//       ),
//     );
//   }
// }
